pkg/ui/Flow.tsx

```diff
@@ -117,29 +116,56 @@ export class Flow<T extends Values> extends Component<Props<T>, State<T>> {
   }
 
   // Handles form submission
-  handleSubmit = (e: MouseEvent | FormEvent) => {
+  handleSubmit = (event: FormEvent<HTMLFormElement> | MouseEvent) => {
     // Prevent all native handlers
-    e.stopPropagation()
-    e.preventDefault()
+    event.stopPropagation()
+    event.preventDefault()
 
     // Prevent double submission!
     if (this.state.isLoading) {
       return Promise.resolve()
     }
 
+    const form = event.currentTarget
+
+    let body: T | undefined
+
+    if (form && form instanceof HTMLFormElement) {
+      const formData = new FormData(form)
+
+      // map the entire form data to JSON for the request body
+      body = Object.fromEntries(formData) as T
+
+      if ("nativeEvent" in event) {
+        // We need the method specified from the name and value of the submit button.
+        // when multiple submit buttons are present, the clicked one's value is used.
+        if ("submitter" in event.nativeEvent) {
+          const method = (
+            event.nativeEvent as unknown as { submitter: HTMLInputElement }
+          ).submitter
+          body = {
+            ...body,
+            ...{ [method.name]: method.value },
+          }
+        }
+      }
```

Comment:
This could be easier to understand with a [custom type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards)
```suggestion
      const hasSubmitter = (evt: any): evt is { submitter: HTMLInputElement } => "submitter" in evt

      // We need the method specified from the name and value of the submit button.
      // when multiple submit buttons are present, the clicked one's value is used.
      if (hasSubmitter(event.nativeEvent)) {
        const method = event.nativeEvent.submitter
        body = {
          ...body,
          ...{ [method.name]: method.value },
        }
      }
```

Reply:
But we don't use this anywhere else? Feels like a bit of more work to add abstraction to such a check, no?

Reply:
I find the code easier to read and understand as it's self documenting and more concise.